### DISCUSSION TITLE
How to use the Android view extensions?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/Decompose#74] [Q&A] [answered]
I noticed that there is a lot of documentation around using Decompose with a declarative UI toolkit and I noticed the Android view extensions module in the project and was wondering if there was an example anywhere on how exactly to use it? 

If I'm understanding the code for `RouterView`, that is a `FrameLayout` which could be inflated and passed a router state through the `children()` function? 

I was also curious how one might animate between children?

**Comment by plusmobileapps** [comment]
I seem to have found my answer digging into the [Counter example](https://github.com/arkivanov/Decompose/blob/master/sample/counter/app-android/src/main/java/com/arkivanov/counter/app/MainActivity.kt#L50).

**Comment by arkivanov** [comment, marked as answer]
The `extensions-android` module is indeed for working with normal Android Views, but it is really really experimental. I will annotated its as experimental to make it explicit. Decompose is primarily designed for declarative UI frameworks, so it does not have things like `onViewCreated`, `viewLifecycle`, etc. And so Android Views have to be attached externally, listen for state changes and update accordingly.

The Counter example indeed contains Android Views sample UI and can be used as reference. With `RouterView` you have to manually add/remove child views, and so you can use e.g. `TransitionManager` for animations.

**Reply by plusmobileapps** [reply]
Ahh I see, that sort of reminds me of how the [Conductor](https://github.com/bluelinelabs/Conductor) library works. I figured Decompose was designed for declarative UI frameworks, was curious on the integration to Android views for work since enterprises like stable things and Compose has just reached beta which still has a bit of time until that hits stable. Is there anything I should be aware of with the android extensions in their experimental state as I play around with them?

**Reply by arkivanov** [reply]
Basically `extensions-android` is experimental because the way how it works is completely new to me. It can remain same and become stable at some point, or the API may change significantly, or it may even go away (if e.g. we discover some blocking issues).

**Reply by arkivanov** [reply]
Recently I annotated everything in this module as experimental, so you will need to opt-in starting with the next release.
